docs/Config.md
# Config

Found in `/src/config/config.json`. Properties not required can be `null`

## MQTT

```json
...
    "mqtt": {
        "host": "192.168.1.2",
        "port": 1883,
        "keepalive": 60,
        "protocol": 5,
        "clientId": "sec_1673108642",

        "topics": {
            "readPower": "power/xxx-xxx-xxx/tele/SENSOR",
            "writeCommand": "solar/xxx/cmd/limit_nonpersistent_relative",
            "inverterStatus": "solar/xxx/status/producing"
        },

        "auth": {
            "username": "my-user",
            "password": "my-password"
        }
    },
...
```

### MQTT Properties

Setup basic mqtt properties

|Req                | Property               | Type              | Default       | Description
|---                | ---                    | ---               |---            |---
| :red_circle:      | `mqtt.host`            | string            |               | hostname or IP address of the remote broker
|                   | `mqtt.port`            | int               | 1883          | network port of the server host to connect to
|                   | `mqtt.keepalive`       | int               | 60            | maximum period in seconds allowed between communications with the broker
|                   | `mqtt.protocol`        | int               | 4             | version of the mqtt protocol to use. `MQTTv31 = 3`, `MQTTv311 = 4`, `MQTTv5 = 5`
|:yellow_circle:    | `mqtt.clientId`        | string            | solar-export-control | mqtt client id to use, required if multiple instances of this program are running
| :red_circle:      | `mqtt.topics`          | object            |               | controls mqtt topics
|                   | `mqtt.auth`            | object            | null          | controls mqtt auth

### MQTT.TOPICS Properties

Setup mqtt topics

|Req                | Property               | Type              | Description
|---                | ---                    | ---               |---
| :red_circle:      | `topics.readPower`     | string            | MQTT-Topic to read current power draw
|                   | `topics.writeCommand`  | string            | MQTT-Topic to write power limit command to
|                   | `topics.inverterStatus`| string            | MQTT-Topic to listens for inverter status updates. This allows to sleep when the inverter is not producing

### MQTT.AUTH Properties

Setup mqtt broker authentication. **Will only be used If `username` is not empty**

|Req                | Property               | Type             | Default       | Description
|---                | ---                    | ---              |---            |---
| :red_circle:      | `auth.username`        | string           |               | set a username for broker authentication
|                   | `auth.password`        | string           | null          | set a password for broker authentication

<br />

---

<br />

## COMMAND

```json
...
    "command": {
        "target": -100,
        "minPower": 24,
        "maxPower": 1200,
        "type": "relative",
        "throttle": 6,
        "hysteresis": 24.0,
        "retransmit": 0,
        "defaultLimit": null
    },
...
```

### COMMAND Properties

Setup how commands will be issued

|Req                | Property                 | Type             | Unit          | Description
|---                | ---                      | ---              |---            |---
| :red_circle:      | `command.target`         | int              | Watt (W)      | power consumption this app will use as target. Typical values are `0` (Zero Export) or `-600` (in Germany "Balkonkraftwerk")
| :red_circle:      | `command.minPower`       | int              | Watt (W)      | the lower power limit the inverter can be set to
| :red_circle:      | `command.maxPower`       | int              | Watt (W)      | the upper power limit the inverter can be set to
| :red_circle:      | `command.type`           | string: "absolute" or "relative"|| controls wether the limit command is absolute in watts (W) or in relative percent of `command.maxPower`
| :red_circle:      | `command.throttle`       | int              | Seconds (s)   | minimum amount of time that must pass after a limit command has been issued before a new one can be issued. Use `0` to disable
| :red_circle:      | `command.hysteresis`     | number           | Watt (W)      | minimum threshold that must been reached after a limit command has been issued before a new one can be issued. Use `0.00` to disable
| :red_circle:      | `command.retransmit`     | int              | Seconds       | time after which `command.hysteresis` is ignored to retransmit the limit command. Useful if commands can get 'lost' on the way to the inverter. Use `0` to disable
|                   | `command.defaultLimit`   | int              | Watt (W)      | default inverter limit which is used during startup as calibration and if `meta.resetInverterLimitOnInactive` is active

<br />

---

<br />

## READING

```json
...
    "reading": {
        "offset": 0,
        "smoothing": "avg",
        "smoothingSampleSize": 8
    },
...
```

### READING Properties

Setup how power reading will be handled

|Req                | Property               | Type             | Default       | Description
|---                | ---                    | ---              |---            |---
|                   | `reading.offset`       | int              | 0             | specifiy an offset in watts (W) to add or subtract
|                   | `reading.smoothing`    | string: "avg" or null| null      | - null: original power reading will be used<br/>- `avg`: average of `reading.smoothingSampleSize` is used<br />Use `avg` to filter short power spikes
|                   | `reading.smoothingSampleSize`| int        | 0             | amount of samples to use for `reading.smoothing` when not `none`

<br />

---

<br />

## META

````json
...
    "meta": {
        "prefix": "solarexportcontrol",
        "resetInverterLimitOnInactive": true,

        "telemetry": {
            "power": true,
            "sample": true,
            "overshoot": true,
            "limit": true,
            "command": true
        },

        "homeAssistantDiscovery": {
            "enabled": true,
            "discoveryPrefix": "homeassistant",
            "id": 1,
            "name": "SEC"
        }
    },
...
````

### META Properties

Setup how this application can be controlled and how it publishes telemetry

|Req                | Property                            | Type          | Description
|---                | ---                                 | ---           |---
| :red_circle:      | `meta.prefix`                       | string        | prefix used for every mqtt topic managed by this application
| :red_circle:      | `meta.resetInverterLimitOnInactive` | bool          | should the inverter limit be reset to max when application is disabled?
| :red_circle:      | `meta.telemtry`                     | object        | manages the information which are published as mqtt topics
| :red_circle:      | `meta.homeAssistantDiscovery`       | object        | manages the home assistant auto discovery

### META.TELEMETRY Properties

Setup which values are published as telemetry

|Req                | Property                            | Type | Unit     | Description
|---                | ---                                 | ---  |---       |---
| :red_circle:      | `telemetry.power`                   | bool | Watt (W) | outputs the raw power value as parsed from `mqtt.topics.readPower`
| :red_circle:      | `telemetry.sample`                  | bool | Watt (W) | outputs the power value after applying `reading.offset` and `reading.smoothing`
| :red_circle:      | `telemetry.overshoot`               | bool | Watt (W) | outputs the difference between the last sample and `command.target`
| :red_circle:      | `telemetry.limit`                   | bool | Watt (W) | outputs the calculated inverter limit
| :red_circle:      | `telemetry.command`                 | bool | Watt (W) or Percent (%) | outputs the last issued inverter limit command as published in `mqtt.topics.writeCommand`. Watt if `command.type` is `absolute`, percent if `relative`

### META.HOMEASSISTANTDISCOVERY

Setup the home assistant integration (auto discovery of telemetry)

|Req                | Property                                 | Type   | Description
|---                | ---                                      | ---    |---
| :red_circle:      | `homeAssistantDiscovery.enabled`         | bool   | enables or disables the integration
| :red_circle:      | `homeAssistantDiscovery.discoveryPrefix` | string | sets home assistant auto discovery topic prefix. Use `homeassistant` unless you have changed this in home assistant
| :red_circle:      | `homeAssistantDiscovery.id`              | int    | used for creating the unique id in home assistant. Only change this if you run multiple instances of this program
|  :red_circle:     | `homeAssistantDiscovery.name`            | string | the name of the device and entites in home assistant

## CUSTOMIZE

```json
...
"customize": {
    "command": {}
}
...
```

### CUSTOMIZE Properties

Specify arbitrary data to pass to `customize.py` functions

|Req                | Property               | Type             | Default        | Description
|---                | ---                    | ---              |---             |---
|                   | `customize.command`    | object           | (empty) object | data passed to `command_to_generic` in `customize.py`


src/config.py
from enum import Enum
import paho.mqtt.client as mqtt
import json


class InverterCommandType(Enum):
    ABSOLUTE = 1
    RELATIVE = 2


class PowerReadingSmoothingType(Enum):
    NONE = 1,
    AVG = 2


class AppConfig:
    def __init__(self, host: str, topic_read_power: str, topic_write_limit: str,
                 inverter_command_throttle: int, inverter_command_min_diff:float, inverter_command_type: InverterCommandType, inverter_max_power: int,
                 power_reading_target:int, power_reading_smoothing: PowerReadingSmoothingType, power_reading_smoothing_sample_size: int,
                 last_will_topic: str | None = None,
                 last_will_payload: str | None = None,
                 last_will_retain: bool | None = None,
                 cred_username: str | None = None,
                 cred_password: str | None = None,
                 port: int | None = None,
                 keepalive: int | None = None,
                 protocol: int | None = None,
                 retain: bool | None = None,
                 client_id: str | None = None,
                 clean_session: bool | None = None,
                 ) -> None:

        self.host: str = host
        self.topic_read_power: str = topic_read_power
        self.topic_write_limit: str = topic_write_limit
        self.inverter_command_throttle: int = inverter_command_throttle
        self.inverter_command_type: InverterCommandType = inverter_command_type
        self.inverter_command_min_diff: float = inverter_command_min_diff
        self.inverter_max_power: int = inverter_max_power
        self.power_reading_target: int = power_reading_target
        self.power_reading_smoothing: PowerReadingSmoothingType = power_reading_smoothing
        self.power_reading_smoothing_sample_size: int = power_reading_smoothing_sample_size
        self.last_will_topic: str | None = last_will_topic
        self.last_will_payload: str | None = last_will_payload
        self.last_will_retain: bool = last_will_retain if last_will_retain is not None else False
        self.cred_username: str | None = cred_username
        self.cred_password: str | None = cred_password
        self.port: int = port if port is not None else 1883
        self.keepalive: int = keepalive if keepalive is not None else 60
        self.protocol: int = protocol if protocol is not None else mqtt.MQTTv311
        self.retain: bool = retain if retain is not None else False
        self.client_id: str = client_id if client_id is not None else "solar-export-control"
        self.clean_session = clean_session if clean_session is not None else True


def config_from_json(path: str) -> AppConfig:
    fs = open(path, "r")
    jf = json.load(fs)

    j_host: str = jf.get("host")
    if j_host is None or type(j_host) is not str or j_host == "":
        raise ValueError("Config: Invalid host")

    j_topic_read_power: str = jf.get("topicReadPower")
    if j_topic_read_power is None or j_topic_read_power == "":
        raise ValueError("Config: Invalid topicReadPower")

    j_topic_write_limit: str = jf.get("topicWriteLimit")
    if j_topic_write_limit is None or j_topic_write_limit == "":
        raise ValueError("Config: Invalid topicWriteLimit")

    j_inverter_command_throttle: int = jf.get("inverterCommandThrottle")
    if j_inverter_command_throttle is None or type(j_inverter_command_throttle) is not int:
        raise ValueError("Config: Invalid inverterCommandThrottle")

    j_inverter_command_type = jf.get("inverterCommandType")
    e_inverter_command_type: InverterCommandType

    if j_inverter_command_type == "absolute":
        e_inverter_command_type = InverterCommandType.ABSOLUTE
    elif j_inverter_command_type == "relative":
        e_inverter_command_type = InverterCommandType.RELATIVE
    else:
        raise ValueError("Config: Invalid inverterCommandType")

    j_inverter_min_diff = jf.get("inverterCommandMinDiff")
    if type(j_inverter_min_diff) is not float or j_inverter_min_diff < 0:
        raise ValueError("Config: Invalid inverterCommandMinDiff")

    j_inverter_max_power = jf.get("inverterMaxPower")
    if type(j_inverter_max_power) is not int or j_inverter_max_power < 0:
        raise ValueError("Config: Invalid inverterMaxPower")


    j_power_reading_target:int = jf.get("powerReadingTarget")
    if type(j_power_reading_target) is not int:
        raise ValueError("Config: Invalid powerReadingTarget")

    j_power_reading_smoothing:str = jf.get("powerReadingSmoothing")
    e_power_reading_smoothing: PowerReadingSmoothingType = PowerReadingSmoothingType.NONE

    if(j_power_reading_smoothing == "avg"):
        e_power_reading_smoothing = PowerReadingSmoothingType.AVG

    j_power_reading_smoothing_sample_size: int = jf.get("powerReadingSmoothingSampleSize")
    if j_power_reading_smoothing_sample_size is None or type(j_power_reading_smoothing_sample_size) is not int or j_power_reading_smoothing_sample_size < 0: 
        j_power_reading_smoothing_sample_size = 0

    j_port: int | None = None
    j_keepalive: int | None = None
    j_protocol: int | None = None
    j_retain: bool | None = None
    j_client_id: str | None = None
    j_clean_session: bool | None = None

    t = jf.get("port")
    if type(t) is int and t > 0:
        j_port = t

    t = jf.get("keepalive")
    if type(t) is int and t > 0:
        j_keepalive = t

    t = jf.get("protocol")
    if type(t) is int:
        j_protocol = t

    t = jf.get("retain")
    if type(t) is bool:
        j_retain = t

    t = jf.get("clientId")
    if type(t) is str:
        j_client_id = t

    t = jf.get("cleanSession")
    if type(t) is bool:
        j_clean_session = t

    j_last_will = jf.get("lastWill")
    j_last_will_topic: str | None = None
    j_last_will_payload: str | None = None
    j_last_will_retain: bool | None = None

    if type(j_last_will) is dict:
        t = j_last_will.get("topic")
        if type(t) is str:
            j_last_will_topic = t

        t = j_last_will.get("payload")
        if type(t) is str:
            j_last_will_payload = t

        t = j_last_will.get("retain")
        if type(t) is bool:
            j_last_will_retain = t

    j_auth = jf.get("auth")
    j_auth_user: str | None = None
    j_auth_pw: str | None = None

    if type(j_auth) is dict:
        t = j_auth.get("username")
        if type(t) is str:
            j_auth_user = t

        t = j_auth.get("password")
        if type(t) is str:
            j_auth_pw = t

    conf = AppConfig(
        host=j_host,
        topic_read_power=j_topic_read_power,
        topic_write_limit=j_topic_write_limit,
        inverter_command_throttle=j_inverter_command_throttle,
        inverter_command_type=e_inverter_command_type,
        inverter_command_min_diff=j_inverter_min_diff,
        inverter_max_power=j_inverter_max_power,
        power_reading_target=j_power_reading_target,
        power_reading_smoothing=e_power_reading_smoothing,
        power_reading_smoothing_sample_size=j_power_reading_smoothing_sample_size,
        last_will_topic=j_last_will_topic,
        last_will_payload=j_last_will_payload,
        last_will_retain=j_last_will_retain,
        cred_username=j_auth_user,
        cred_password=j_auth_pw,
        port=j_port,
        keepalive=j_keepalive,
        protocol=j_protocol,
        retain=j_retain,
        client_id=j_client_id,
        clean_session=j_clean_session
    )

    return conf
